typed-racket-lib/typed-racket/typecheck/tc-lambda-unit.rkt

```diff
@@ -117,71 +112,110 @@
 ;; rest-id: The identifier of the rest arg, or #f for no rest arg
 ;; body: The body of the lambda to typecheck.
 ;; arg-tys: The expected positional argument types.
-;; rst: #f, expected rest arg Type, or expected RestDots
+;; rst: #f, expected rest arg Rest, or expected RestDots
 ;; ret-ty: The expected type of the body of the lambda.
 (define/cond-contract (check-clause arg-list rest-id body arg-tys rst ret-ty)
   ((listof identifier?)
-   (or/c #f identifier?) syntax? (listof Type?) (or/c #f Type? RestDots?)
+   (or/c #f identifier?) syntax? (listof Type?) (or/c #f Rest? RestDots?)
    tc-results/c
    . -> .
    Arrow?)
   (let* ([arg-len (length arg-list)]
-         [tys-len (length arg-tys)]
+         [arg-tys-len (length arg-tys)]
+         [extra-arg-count (- arg-len arg-tys-len)]
          [arg-types
-          (if (andmap type-annotation arg-list)
-              (get-types arg-list #:default Univ)
-              (cond
-                [(= arg-len tys-len) arg-tys]
-                [(< arg-len tys-len) (take arg-tys arg-len)]
-                [(> arg-len tys-len)
-                 (append arg-tys
-                         (map (if (Type? rst)
-                                  (λ _ rst)
-                                  (λ _ -Bottom))
-                              (drop arg-list tys-len)))]))])
+          (cond
+            [(andmap type-annotation arg-list)
+             (get-types arg-list #:default Univ)]
+            [(eqv? 0 extra-arg-count) arg-tys]
+            [(negative? extra-arg-count) (take arg-tys arg-len)]
+            [else
+             (define tail-tys (match rst
+                                [(Rest: rst-tys)
+                                 (define rst-len (length rst-tys))
+                                 (for/list ([idx (in-range extra-arg-count)])
+                                   (list-ref rst-tys (remainder idx rst-len)))]
```

Comment:
why not:
```
(for/list ([_ (in-range extra-arg-count]
           [ty (in-cycle rst-tys)])
  ty)
```

Reply:
good point! But `in-cycle` super slow dynamic dispatch yuckiness... so I wrote an `in-list-cycle` that is __significantly__ faster and achieves the same purpose. Now it's efficient and simple =)

The only sequence we're ever working with is lists anyways, so that should be a potentially useful macro elsewhere as well.